build.gradle

```diff
@@ -139,6 +164,72 @@ allprojects {
     }
 }
 
+def log4jVersion = '2.22.1'
+
+configurations {
+    distOnly
+}
+
+dependencies {
+    distOnly "org.apache.logging.log4j:log4j-core:${log4jVersion}"
+    distOnly "org.apache.logging.log4j:log4j-slf4j2-impl:${log4jVersion}"
+}
+
+tasks.register('dist') {
+    group = JavaBasePlugin.BUILD_TASK_NAME
+    description "Creates the distribution directory"
+    def includedProjects =
+            ["eventmesh-common",
+             "eventmesh-meta:eventmesh-meta-api",
+             "eventmesh-metrics-plugin:eventmesh-metrics-api",
+             "eventmesh-protocol-plugin:eventmesh-protocol-api",
+             "eventmesh-retry:eventmesh-retry-api",
+             "eventmesh-runtime",
+             "eventmesh-security-plugin:eventmesh-security-api",
+             "eventmesh-spi",
+             "eventmesh-starter",
+             "eventmesh-storage-plugin:eventmesh-storage-api",
+             "eventmesh-trace-plugin:eventmesh-trace-api",
+             "eventmesh-webhook:eventmesh-webhook-api",
+             "eventmesh-webhook:eventmesh-webhook-admin",
+             "eventmesh-webhook:eventmesh-webhook-receive"]
+    includedProjects.each { dependsOn("${it}:jar") }
+    outputs.dirs('dist/apps', 'dist/bin', 'dist/conf', 'dist/lib', 'dist/licenses')
```

Comment:
May I ask what these two lines are for? How did you come up with the idea of creating these two lines, did the IDE prompt you?

I didn't understand the usage of `dependsOn("${it}:jar")`.

It seems `outputs.dirs` equals to `mkdirs()`, but I haven't found much documents and examples proving that.

(A little too cautious with old but working scripts

Reply:
The first line should inform Gradle that the `:dist` tasks depends on all the `jar` tasks of the projects in includes `:eventmesh-common:jar`, `:eventmesh-meta:eventmesh-meta-api:jar`, etc.

The second line should inform Gradle that the `:dist` task creates and fills those directories. It should help Gradle in determining if `:dist` needs to run or is up-to-date (cf. [Benefits of declaring task inputs and outputs](https://docs.gradle.org/current/userguide/incremental_build.html#sec:task_input_output_side_effects)). Disclaimer: I mostly use Maven in my day job, so my Gradle knowledge is somehow limited.

Reply:
Thank you! This explanation is very detailed.

>The first line should inform Gradle that the `:dist` tasks depends on all the `jar` tasks of the projects in includes `:eventmesh-common:jar`, `:eventmesh-meta:eventmesh-meta-api:jar`, etc.

It seems that this line didn't work, according to https://github.com/apache/eventmesh/pull/4719#discussion_r1554553497.

>The second line should inform Gradle that the `:dist` task creates and fills those directories.

For the `dist` task, I don't think we need to declare its output folder. Because when we execute the `dist` task, we definitely want it to repack a copy. Besides, the `dist` task is a terminal task, and there won't be any more tasks depending on it.

Reply:
> > The first line should inform Gradle that the `:dist` tasks depends on all the `jar` tasks of the projects in includes `:eventmesh-common:jar`, `:eventmesh-meta:eventmesh-meta-api:jar`, etc.
> 
> It seems that this line didn't work, according to [#4719 (comment)](https://github.com/apache/eventmesh/pull/4719#discussion_r1554553497).

Fixed in 8d5cab712b496b5166980d9f103075eb8f2c31ca, the problem was the dependencies of the `installPlugin` task.

> > The second line should inform Gradle that the `:dist` task creates and fills those directories.
> 
> For the `dist` task, I don't think we need to declare its output folder. Because when we execute the `dist` task, we definitely want it to repack a copy. Besides, the `dist` task is a terminal task, and there won't be any more tasks depending on it.

Doesn't `installPlugin` and `tar` depend on `dist`? I can replace the `outputs.dirs` with `mkdirs` if you prefer.

Reply:
`installPlugin` is a seperate task (according to https://eventmesh.apache.org/docs/instruction/runtime/#24-package-plugins), which can bring more flexibility in plugin file loading. So I think `installPlugin` doesn't need to depend on `dist`, and `dist` should depend on `jar`.

In terms of task scripts, I think there isn't any logic that needs to be changed, it's better to minimize logic changes ~ the grammar change is totally ok.

Reply:
I restored the dependency of `:dist` on all `:*:jar` tasks.